async_simple/coro/Collect.h

```diff
@@ -122,6 +128,83 @@ struct CollectAnyAwaiter {
 };
 
+template <template <typename> typename LazyType, typename... Ts>
+struct CollectAnyVariadicAwaiter {
+    using ResultType = std::variant<Try<Ts>...>;
+    using InputType = std::tuple<LazyType<Ts>...>;
+
+    CollectAnyVariadicAwaiter(LazyType<Ts>&&... inputs)
+        : _input(std::make_unique<InputType>(std::move(inputs)...)),
+          _result(nullptr) {}
+
+    CollectAnyVariadicAwaiter(const CollectAnyVariadicAwaiter&) = delete;
+    CollectAnyVariadicAwaiter& operator=(const CollectAnyVariadicAwaiter&) =
+        delete;
+    CollectAnyVariadicAwaiter(CollectAnyVariadicAwaiter&& other)
+        : _input(std::move(other._input)), _result(std::move(other._result)) {}
+
+    bool await_ready() const noexcept {
+        return std::tuple_size<InputType>() == 0 ||
+               (_result && _result->has_value());
+    }
+
+    template <size_t... index>
+    void await_suspend_impl(std::index_sequence<index...>,
+                            std::coroutine_handle<> continuation) {
+        auto promise_type =
+            std::coroutine_handle<LazyPromiseBase>::from_address(
+                continuation.address())
+                .promise();
+        auto executor = promise_type._executor;
+
+        auto input = std::move(_input);
+        // Make local copies to shared_ptr to avoid deleting objects too early
+        // if any coroutine finishes before this function.
+        auto result = std::make_shared<std::optional<ResultType>>();
+        auto event =
+            std::make_shared<detail::CountEvent>(std::tuple_size<InputType>());
+
+        _result = result;
+
+        (
+            [&]() {
+                if (result->has_value()) {
+                    return;
+                }
+                if (!std::get<index>(*input)._coro.promise()._executor) {
+                    std::get<index>(*input)._coro.promise()._executor =
+                        executor;
+                }
+                std::get<index>(*input).start(
+                    [r = result, c = continuation,
+                     e = event](std::variant_alternative_t<index, ResultType>&&
+                                    res) mutable {
+                        assert(e != nullptr);
+                        auto count = e->downCount();
+                        if (count == std::tuple_size<InputType>() + 1) {
+                            *r = ResultType{std::in_place_index_t<index>(),
+                                            std::move(res)};
+                            c.resume();
+                        }
+                    });
+            }(),
+            ...);
+    }
+
+    void await_suspend(std::coroutine_handle<> continuation) {
+        await_suspend_impl(std::make_index_sequence<sizeof...(Ts)>{},
+                           std::move(continuation));
+    }
+
+    auto await_resume() {
+        assert(_result != nullptr);
+        return std::move(_result->value());
+    }
+
+    std::unique_ptr<std::tuple<LazyType<Ts>...>> _input;
```

Comment:
Is it OK to make `_input` as `std::tuple<LazyType<Ts>...>`?

Reply:
move may be expensive for tuple